src/couch_stats/src/csrt_query.erl

```diff
@@ -0,0 +1,173 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(csrt_query).
+
+-include_lib("stdlib/include/ms_transform.hrl").
+-include_lib("couch_stats_resource_tracker.hrl").
+
+%% aggregate query api
+-export([
+    active/0,
+    active/1,
+    active_coordinators/0,
+    active_coordinators/1,
+    active_workers/0,
+    active_workers/1,
+    count_by/1,
+    find_by_nonce/1,
+    find_by_pid/1,
+    find_by_pidref/1,
+    find_workers_by_pidref/1,
+    group_by/2,
+    group_by/3,
+    sorted/1,
+    sorted_by/1,
+    sorted_by/2,
+    sorted_by/3
+]).
+
+%%
+%% Aggregate query API
+%%
+
+active() ->
+    active_int(all).
+
+active_coordinators() ->
+    active_int(coordinators).
+
+active_workers() ->
+    active_int(workers).
+
+%% active_json() or active(json)?
+active(json) ->
+    to_json_list(active_int(all)).
+
+active_coordinators(json) ->
+    to_json_list(active_int(coordinators)).
+
+active_workers(json) ->
+    to_json_list(active_int(workers)).
+
+active_int(coordinators) ->
+    select_by_type(coordinators);
+active_int(workers) ->
+    select_by_type(workers);
+active_int(all) ->
+    select_by_type(all).
+
+select_by_type(coordinators) ->
+    ets:select(?MODULE, ets:fun2ms(fun(#rctx{type = #coordinator{}} = R) -> R end));
+select_by_type(workers) ->
+    ets:select(?MODULE, ets:fun2ms(fun(#rctx{type = #rpc_worker{}} = R) -> R end));
+select_by_type(all) ->
+    ets:tab2list(?MODULE).
+
+find_by_nonce(Nonce) ->
+    %%ets:match_object(?MODULE, ets:fun2ms(fun(#rctx{nonce = Nonce1} = R) when Nonce =:= Nonce1 -> R end)).
+    [R || R <- ets:match_object(?MODULE, #rctx{nonce = Nonce})].
+
+find_by_pid(Pid) ->
+    %%[R || #rctx{} = R <- ets:match_object(?MODULE, #rctx{pid_ref={Pid, '_'}, _ = '_'})].
+    [R || R <- ets:match_object(?MODULE, #rctx{pid_ref = {Pid, '_'}})].
```

Comment:
This is a second place where we use internal representation. One way to improve it would be:

```
%% couch_stats_resource_tracker.hrl
-define(resource_id_pid(Pid), {Pid, '_'}).

%% csrt_query.erl
[R || R <- ets:match_object(?MODULE, #rctx{pid_ref = ?resource_id_pid(Pid)})].
```

Reply:
If I understand the code correctly there is a mistake in here. I couldn't find an ets table named `csrt_query`. Most likely this should be `csrt_server` instead of `?MODULE`. If I am right then you can just move this matcher implementation into `csrt_server.erl` and have a dispatch here.

```erlang
%% csrt_query.erl
find_by_pid(Pid) ->
     csrt_server:find_by_pid(Pid).

%% csrt_server.erl
find_by_pid(Pid) ->
    [R || R <- ets:match_object(?MODULE, #rctx{pid_ref = {Pid, '_'}})].
```